src/server/controllers/votes.js

```diff
@@ -1,17 +1,93 @@
 import { respondWithSuccess } from '~/server/helpers/respond';
 import dispatch from '~/server/services/dispatcher';
-import requestGraph from '~/common/services/subgraph';
 import Vote from '~/server/models/vote';
-import Answer from '~/server/models/answer';
+import { QuestionHasManyAnswers } from '~/server/database/associations';
+import Question from '~/server/models/question';
 import baseController from '~/server/controllers';
+import {
+  filterResponseFieldsAll,
+  filterResponseFields,
+} from '~/server/controllers';
+
+const customFilter = (req, data, options) => {
+  let combined = combineWithGraph(
+    req.locals.graphData,
+    data.answers,
+    'chainId',
+  );
+  if (req.locals && req.locals.user) {
+    data.set('answers', combined, {
+      raw: true,
+    });
+    return filterResponseFieldsAll(req, data, options);
+  }
+  const topThree = findTopThree(combined, 'votePower');
+  combined = combined.map(datum => {
+    if (topThree.includes(parseInt(datum['votePower']))) {
+      const optionsOverride = {
+        ...options,
+        fields: options.fields.concat(options.fieldsProtected),
+      };
+      return filterResponseFields(req, datum, optionsOverride);
+    }
+    return filterResponseFields(req, datum, options);
+  });
+  data.set('answers', combined, {
+    raw: true,
+  });
+  return data;
+};
 
 const options = {
-  model: Answer,
-  fields: ['id', 'votes', 'voteTokens', 'votePower', 'question'],
-  thresholdFields: ['type', 'artworkId', 'propertyId'],
-  threshold: 3,
-  thresholdKey: 'votePower',
-  graphMatchingKey: 'chainId',
+  model: Question,
+  fields: [
+    'title',
+    'address',
+    'answers',
+    'votes',
+    'voteTokens',
+    'votePower',
+    'question',
+  ],
+  fieldsProtected: ['id', 'chainId', 'type', 'artworkId', 'propertyId'],
+  include: [QuestionHasManyAnswers],
+  associations: [
+    {
+      association: QuestionHasManyAnswers,
+      destroyCascade: true,
+    },
+  ],
+  customFilter,
+};
+
+const combineWithGraph = (graphData, apiData, matchingKey) => {
+  const combined = graphData.map(i => {
+    const datum = apiData.find(d => d[matchingKey] === i.id);
+    if (datum) return Object.assign(datum.dataValues, i);
+    return i;
+  });
+  return combined;
+};
+
+const findTopThree = (array, matchingKey) => {
+  let first = -1;
+  let second = -1;
+  let third = -1;
+  let results = array.map(item => parseInt(item[matchingKey]));
+  results = new Set(results);
+  results.forEach(item => {
+    if (item > first) {
+      third = second;
+      second = first;
+      first = item;
+    } else if (item > second) {
+      third = second;
+      second = item;
+    } else if (item > third) {
+      third = item;
+    }
+  });
+  return [first, second, third];
```

Comment:
Maybe there is a reason for writing the sorting like this but would this work as well?

```js
const findTopThree = (array, matchingKey) => {
  return array
     .map((item) => parseInt(item[matchingKey]))
     .sort((itemA, itemB) => itemA - itemB)
     .slice(0, 3);
};
```

Reply:
Pretty sure it would work the same, as long as cast to a set first, but I found some post about how this is the most efficient way if you don't need to sort the whole array :joy: